lib/Serialization/Deserialization.cpp

```diff
@@ -4148,7 +4163,21 @@ Expected<Type> ModuleFile::getTypeChecked(TypeID TID) {
     decls_block::BoundGenericTypeLayout::readRecord(scratch, declID, parentID,
                                                     rawArgumentIDs);
 
-    auto nominal = cast<NominalTypeDecl>(getDecl(declID));
+    auto *decl = getDecl(declID);
+    if (auto *alias = dyn_cast<TypeAliasDecl>(decl)) {
+      // FIXME: We should substitute through the alias rather than just relying
+      // on it forwarding arguments perfectly to the nominal.
+      auto *nominal = alias->getUnderlyingTypeLoc().getType()->getAnyNominal();
+      assert(nominal && "alias does not resolve to a nominal type");
+      // This is not a very good test because of `typealias Foo<X> = Bar<X?>`,
+      // but it's better than nothing.
```

Comment:
Oh, of course, the canonical interface types should also match up. Then we're back to your X/Y, Y/X case. Thanks.